syntax = "proto3";

message SearchRequest1 {
  // The name of the item (leave blank to display all items)
  string item_name = 1;

  // The category of the item (ANY will display items with all categories)
  enum ItemCategory {
    ELECTRONICS = 0;
    FASHION = 1;
    OTHERS = 2;
    ANY = 3;
  }
  ItemCategory item_category = 2;
}

message Item {
  // Information about the item
  float price = 2;
  string product_name = 3;
  string category = 4;
  string description = 5;
  int32 quantity = 6;
  float rating = 7;
}

message SearchResponse1 {
  // List of items matching the search criteria
  repeated Item items = 1;
}

message BuyRequest1 {
  string item_id = 1;
  int32 quantity = 2;
  string buyer_address = 3;
}

message BuyResponse1 {
  enum Status {
    SUCCESS = 0;
    FAILED = 1;
  }
  Status status = 1;
}

message AddToWishListRequest1 {
  string item_id = 1;
  string buyer_address = 2;
}

message AddToWishListResponse1 {
  enum Status {
    SUCCESS = 0;
    FAILED = 1;
  }

  Status status = 1;
}

message RateItemRequest1 {
  string item_id = 1;
  string buyer_address = 2;
  int32 rating = 3;
}

message RateItemResponse1 {
  enum Status {
    SUCCESS = 0;
    FAILED = 1;
  }

  Status status = 1;
}

service BuyerService {
    rpc SearchItem(SearchRequest1) returns (SearchResponse1);
    rpc BuyItem(BuyRequest1) returns (BuyResponse1);
    rpc AddToWishList(AddToWishListRequest1) returns (AddToWishListResponse1);
    rpc RateItem(RateItemRequest1) returns (RateItemResponse1);
}
